docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+
+Shoot OpenVPN clients connect to the correct OpenVPN Server using the http proxy feature provided by OpenVPN. A configuration is added to the envoy proxy to detect http proxy requests and forward the connection attempt to the correct OpenVPN server.
+
+The restriction regarding the `192.268.123.0/24` network range in the current VPN solution still applies to this proposal. No other restrictions are introduced.
+
+### Performance and Scalability
+
+We do expect performance and throughput to be slightly lower compared to the existing solution. This is because the OpenVPN server acts as an additional hop and must decrypt and re-encrypt traffic that passes through. As there are no low latency, or high thoughput requirements for this connection we do not assume this to be an issue.
+
+### Availablity and Failure Scenarios
+
+This solution re-uses multiple instances of the envoy component used for the kube-apiserver endpoints. We assume that the availability for kube-apiservers is good enough for the cluster VPN as well. 
+
+The OpenVPN client- and server pods are singleton pods in this approach and therefore are affected by potential failures and during cluster-, and control plane updates. Potentiall outages are only restricted to single shoot clusters and are comparable to the situation with the existing solution today.
+
+### Feature Gates and Migration Strategy
+
+We have introduced a gardenlet feature gate `ReversedVPN`. If both, SNI and `ReversedVPN` are enabled the proposed solution is automatically enabled for all shoot clusters hosted by the seed. All existing shoot clusters will automatically be migrated during the next reconciliation. We have also added a shoot annotation `alpha.featuregates.shoot.gardener.cloud/reversed-vpn` which can override the feature gate to enable or disable the solution for individual clusters.
+
+### Security Review
+
+The change in the VPN solution will potentially open up new attack vectors. We will perform a thorough analysis outside of this document.
+
+## Alternatives
+
+We have done a detailed investigation and implementation of a reversed VPN based on WireGuard. While we belive that it is technically feasible and superior to the approach presented above there are some concerns with regards to scalability, and high availability. As the WireGuard scenario based on kubelink is relevant for other use cases we continue to improve this implementation and address the concerns but we concede that this might not be on time for the cluster VPN. We nevertheless keep the implementation and provide an outline as part of this proposal.
```

Comment:
```suggestion
We have done a detailed investigation and implementation of a reversed VPN based on WireGuard. While we believe that it is technically feasible and superior to the approach presented above there are some concerns with regards to scalability, and high availability. As the WireGuard scenario based on kubelink is relevant for other use cases we continue to improve this implementation and address the concerns but we concede that this might not be on time for the cluster VPN. Nevertheless, we keep the implementation and provide an outline as part of this proposal.
```

Reply:
You talk about kubelink without having introduced it. This makes this block quasi useless for people which are not familiar with it.

Reply:
Well, maybe it makes more sense to put this "wireguard" idea into a document in the `github.com/gardener/vpn{2}` repository and only link it from here. It is a completely different proposal which won't be implemented, and now at least the half of the proposal contains information about a completely different alternative. Can we rather link it so that only people interested in it can look it at, but focus this document on the desired/planned solution?

Reply:
The WireGuard option is something quite different from what we have in the vpn2 repository and does not belong there. As we invested significant effort, both implementing and discussing the WireGuard solution it well justifies a mention in the proposal as an alternative (which by the way is also compliant with the GEP template).